src/Compilers/Core/Portable/Symbols/IMethodSymbol.cs

```diff
@@ -214,6 +215,19 @@ public interface IMethodSymbol : ISymbol
         /// </summary>
         ImmutableArray<AttributeData> GetReturnTypeAttributes();
 
+        /// <summary>
+        /// The calling convention enum of the method method symbol. If this value is <see cref="SignatureCallingConvention.Unmanaged"/>,
+        /// then <see cref="CallingConventionTypes"/> may have additional calling convention types that are considered part of the
+        /// calling convention of the method.
+        /// </summary>
+        SignatureCallingConvention CallingConvention { get; }
+
+        /// <summary>
+        /// Modifier types that are considered part of the calling convention of this method. If <see cref="CallingConvention"/> is not
+        /// <see cref="SignatureCallingConvention.Unmanaged"/>, then this will be an empty array.
+        /// </summary>
+        ImmutableArray<INamedTypeSymbol> CallingConventionTypes { get; }
```

Comment:
>CallingConventionTypes [](start = 41, length = 22)

We should probably make it clear that this is relevant only for ```IFunctionPointerTypeSymbol.Signature```, or perhaps move it to ```IFunctionPointerTypeSymbol```.

Reply:
I didn't because my thought was that when we implement `UnmanagedCallersOnly`, these properties will become relevant to them.

---
In reply to: [463758407](https://github.com/dotnet/roslyn/pull/46467#discussion_r463758407) [](ancestors = 463758407)

Reply:
> when we implement UnmanagedCallersOnly, these properties will become relevant to regular methods.

UnmanagedCallersOnly is an attribute, right? Could you remind what could be specified through it?

---
In reply to: [463765922](https://github.com/dotnet/roslyn/pull/46467#discussion_r463765922) [](ancestors = 463765922,463758407)

Reply:
> UnmanagedCallersOnly is an attribute, right? Could you remind what could be specified through it?

Right. It will tell the JIT, when emitting the method, to emit a stub header that uses an umanaged calling convention. It'll be an error to call such a function directly in C#, you must take a function pointer to it with the correct calling convention. ie, `delegate* unmanaged[Cdecl]<void> = &MethodWithUnmanagedCallersOnlyAndCallConvCdeclSpecified;`.

---
In reply to: [463768020](https://github.com/dotnet/roslyn/pull/46467#discussion_r463768020) [](ancestors = 463768020,463765922,463758407)

Reply:
Is the array of types the only thing that can be specified by the attribute application?


---
In reply to: [463770629](https://github.com/dotnet/roslyn/pull/46467#discussion_r463770629) [](ancestors = 463770629,463768020,463765922,463758407)

Reply:
Correct. How we'll interpret that array is specified by the proposal: [https://github.com/dotnet/csharplang/blob/master/proposals/csharp-9.0/function-pointers.md#systemruntimeinteropservicesunmanagedcallersonlyattribute](https://github.com/dotnet/csharplang/blob/master/proposals/csharp-9.0/function-pointers.md#systemruntimeinteropservicesunmanagedcallersonlyattribute)

---
In reply to: [463773248](https://github.com/dotnet/roslyn/pull/46467#discussion_r463773248) [](ancestors = 463773248,463770629,463768020,463765922,463758407)

Reply:
Do we really want to modify CallingConvention to Unmanaged in that case? I would expect CallingConvention  to reflect the bit from metadata signature and the attribute is just an information on the side that affects how the method can be used in C#.

---
In reply to: [463774522](https://github.com/dotnet/roslyn/pull/46467#discussion_r463774522) [](ancestors = 463774522,463773248,463770629,463768020,463765922,463758407)

Reply:
My thought is that, from a public API and runtime perspective, the bit from the signature isn't correct. The runtime will throw an exception if managed code attempts to call a method attributed with UnmanagedCallersOnly directly. As a usage example, let's say the IDE wants to add a refactoring when you attempt to call one of these directly, to make it a function pointer invocation instead. They would be able to just grab the calling convention and relevant types from the `IMethodSymbol`, rather than having to try and do the same interpretation of the attribute we'll have to do. I also believe implementing it like this will make our side easier as well, particularly when looking at conversions. We'll be able to just compare MethodSymbol calling conventions directly, and not concern ourselves with whether one side is an resolved address of method group or an actual function pointer type.

---
In reply to: [463776322](https://github.com/dotnet/roslyn/pull/46467#discussion_r463776322) [](ancestors = 463776322,463774522,463773248,463770629,463768020,463765922,463758407)

Reply:
I cannot say that I agree with this. We certainly should expose an API to get attribute information in a convenient way, as we do for many other well-known attributes. But I think that these properties should reflect what is actually in metadata.

---
In reply to: [463808040](https://github.com/dotnet/roslyn/pull/46467#discussion_r463808040) [](ancestors = 463808040,463776322,463774522,463773248,463770629,463768020,463765922,463758407)

Reply:
It also feels wrong to me to mix two different sources of information (attribute, metadata signature) in the same API

---
In reply to: [463842919](https://github.com/dotnet/roslyn/pull/46467#discussion_r463842919) [](ancestors = 463842919,463808040,463776322,463774522,463773248,463770629,463768020,463765922,463758407)

Reply:
How would you feel about making the public API this shape, and internally having two separate properties, something like `EffectiveCallingConvention` plus the existing one (maybe renamed to `MetadataCallingConvention`)? I do feel that having a mechanism for asking about the effective calling convention will be useful in implementing `UnmanagedCallersOnly`, and I also do feel that for our public API, we should make a consistent form of getting at the convention elements. We chose to use an `IMethodSymbol` to represent the signature because it represents exactly the same set of data as a normal method signature, and I think that argument applies to calling conventions as well.

---
In reply to: [463843942](https://github.com/dotnet/roslyn/pull/46467#discussion_r463843942) [](ancestors = 463843942,463842919,463808040,463776322,463774522,463773248,463770629,463768020,463765922,463758407)

Reply:
> We chose to use an IMethodSymbol to represent the signature because it represents exactly the same set of data as a normal method signature, and I think that argument applies to calling conventions as well.

I don't think this is the case here. Methods do not have CallingConventionTypes in my mind. This is where function pointers diverge. I believe that information from an attribute Is a separate piece of information and should be exposed differently, as a side information, and it shouldn't affect calling convention of the method. Internally we have more flexibility, but here we are talking about public API.

---
In reply to: [463860233](https://github.com/dotnet/roslyn/pull/46467#discussion_r463860233) [](ancestors = 463860233,463843942,463842919,463808040,463776322,463774522,463773248,463770629,463768020,463765922,463758407)

Reply:
I suggest we start with a minimal set of APIs without any unification helpers, etc. And we can enhance API based on specific user's requests/feedback.

---
In reply to: [463896497](https://github.com/dotnet/roslyn/pull/46467#discussion_r463896497) [](ancestors = 463896497,463860233,463843942,463842919,463808040,463776322,463774522,463773248,463770629,463768020,463765922,463758407)

Reply:
On the second thought, I would feel more comfortable about the current API shape if we remove stated dependency between two properties: ```If <see cref="CallingConvention"/> is not
 <see cref="SignatureCallingConvention.Unmanaged"/>, then this will be an empty array.``` I.e. this would be true for FPT.Signature, but not for a regular method. And we document that the source of information is different for these two cases.

---
In reply to: [463897224](https://github.com/dotnet/roslyn/pull/46467#discussion_r463897224) [](ancestors = 463897224,463896497,463860233,463843942,463842919,463808040,463776322,463774522,463773248,463770629,463768020,463765922,463758407)